### ISSUE TITLE
List free functions alongside methods in autocompletion

### ISSUE THREAD

**Issue by Marko-Vujnovic** [clangd/clangd#893]
This proposal is not to be confused with the existing feature request (and its issue #850) proposing the addition of postfix auto completion for snippets - which deals mostly with control flow constructs. They are quite unrelated, even though the name might sound similar. 

Motivation: less coupling, fewer circular dependencies and better incremental build times can be achieved by writing free functions instead of methods where possible. This practice, however, comes with significant disadvantages in the form of worse autocompletion and the requirement of previous familiarity with the codebase in order to even be aware of the existence of all potentially relevant free functions importable from somewhere, whereas with methods all relevant methods can easily be found organically just by using the autocompletion feature. The disadvantages of using free functions are very much solvable by tooling. Specifically, the language server can, when listing methods, also list all global functions which take the object by (const) reference as the first argument of the function. When a global function is picked from the autocompletion menu the following transformation should occur:
obj.func() -> func(obj)
Thereafter, the cursor should be placed either after the closing parenthesis or where the next function argument is expected if the function in question takes any additional arguments.

This feature is already present in Jetbrains’ CLion IDE where it’s presumably implemented within their private fork of clangd, and it goes by the name postfix autocompletion for free functions. If you try out the CLion feature you can see that it works nicely with the exception that auto importing of such functions doesn't work. My proposal also includes automatic inclusion of the file containing the free function. In other words, functions should be suggested by auto completion even if they aren't already included and should get automatically included upon selection.

While not mandatory for the initial MVP implementation, this should ideally also work with templates.
void func(auto& obj);

**Comment by HighCommander4** [comment]
Agreed, this would be a nice feature.

One thing to note: while much of the feature requires changes at the same code layers as described in https://github.com/clangd/clangd/issues/850#issuecomment-940705224 (and indeed, reading through that comment is a good starting point if someone would like to take a stab at this), the ability to offer proposals for free functions defined in files which have not yet been included would also require changes to clangd's index, which is not currently structured to efficiently answer queries of the form "list all functions whose first argument matches this type".

**Comment by sam-mccall** [comment]
+1, this would be great. There are challenges...

> the ability to offer proposals for free functions defined in files which have not yet been included would also require changes to clangd's index

I think this is required for the feature to work at all. Even for headers that *are* included, free functions are looked up for code completion from the index, not the AST. The AST doesn't provide an efficient way to find "all functions that match this fuzzy query" (and the preamble doesn't have a way to deserialize only those, either).

---

Apart from getting this to trigger, there's going to be a challenge in trying not to trigger it too much, and in ranking.
Consider `std::string`: our internal codebase probably has around 1M functions that take a string as the first parameter.
Even if we work out how to order these, there's a large cost to "filling up the page" with results that aren't an integral part of `string`, namely it becomes difficult to use the completion window to browse the API because you never know when you're done.
I suspect the scope needs to be carefully limited. Files that are defined next to the type are good candidates (though your point about isolating dependencies is valid!). Results from non-included headers may bring more noise than signal...

---

> While not mandatory for the initial MVP implementation, this should ideally also work with templates.

I don't see a way to make this work well. Once you've defined `void foo(auto&);` somewhere in your project, we don't want **every** member completion on every object to show `foo()` as an option. There may be hundreds of these functions!
We certainly can't afford to instantiate the template to see if it makes sense.
I think this feature would need to be restricted to functions that explicitly name the target type.

**Comment by Marko-Vujnovic** [comment]
> I think this feature would need to be restricted to functions that explicitly name the target type.
Not necessarily. Would you agree that it'd be feasible if the scope of the feature is reduced to only supporting templates where a constraint is present?
void send(Serializable auto& obj);
> Results from non-included headers may bring more noise than signal.
One of the main purposes of autocompletion is being immediately productive in an unfamiliar codebase. If I already know the exact function I need and in which specific file it resides, then I don't need autocompletion at all just to save me from typing a few extra characters by hand. The scenario where this feature would be most useful is when a person finds themselves in a new codebase where they have no clue what free functions even exist that can operate on the type at hand. Without auto importing this feature is close to useless to me.
> Files that are defined next to the type are good candidates (though your point about isolating dependencies is valid!). 
I'd like to separate some of the free functions into a separate library that the users of the main library can optionally also include, but only if they need the functionality defined therein. I think we can please everybody by making the behavior without this limitation something that a person can opt into if they want. The default behavior can be as you outlined, or similar, but I think having this opt-in switch is crucial for some users like myself.

**Comment by HighCommander4** [comment]
> > I think this feature would need to be restricted to functions that explicitly name the target type.
> 
> Not necessarily. Would you agree that it'd be feasible if the scope of the feature is reduced to only supporting templates where a constraint is present? void send(Serializable auto& obj);

I think that's definitely worth exploring. Whether it ends up being feasible will likely depend on how the performance shakes out in practice, i.e. whether checking such constraints on the fly (possibly with some amount of caching) can be done without too much impact to the latency of completion.

**Comment by sam-mccall** [comment]
Yeah, concept-constrained params sound good if they can be made to work.

There's a performance question (concept checking is template instantiation, which is often multiple milliseconds, check `clang -ftime-trace`).
There's also an issue with the AST: the definition of `Serializable` may not be available where you're trying to complete (it's visible from `send()`, but if that might not be imported...)
Finally some concepts like `Copyable` aren't much more restrictive than plain `auto` so there's still a spam issue.

I suspect the best answer to all these is some cache/index of which types we've seen checked against which concepts as part of the normal build. (This is probably also key making the "expected type" part of code complete ranking work with concepts.)

> Without auto importing this feature is close to useless to me.

I see what you're saying, but you didn't really address my concern. If it's true that results from included headers are too obvious to be useful, and it's also true that codebase-wide results are too spammy to be useful, the conclusion would be that we can't make this feature useful at all!
I hope it's not true, and don't want to be discouraging - I hope we can find heuristics to identify associated/relevant functions.

> I think we can please everybody by making the behavior without this limitation something that a person can opt into if they want.

We can definitely have options to tweak the behavior here. (And I suspect purists will want an option to turn it off entirely).
I'd caution that we should find a part of the design space that "just works" well enough to turn on by default in some form.
If the feature is **entirely** opt-in then it'll never be used by more than a few percent, which makes it hard to justify the complexity, extra ram for the index etc that it entails.

**Comment by calestialgem** [comment]
Is there any will to implement this? Seems very useful to me.

**Comment by HighCommander4** [comment]
> Is there any will to implement this?

Reviewing the discussion so far, this seems like a useful feature to have but there are open questions about performance and noisy-ness of results.

A good next step to making progress on this is to prototype an implementation and evaluate these aspects (performance and noise) on one or more real-life codebases.

If someone is interested in picking up this project, I'm happy to provide guidance. (I've already written general pointers for extending clangd's code completion in [this comment](https://github.com/clangd/clangd/issues/850#issuecomment-940705224), but I'm happy to go into more specifics related to this issue.)
